### ISSUE TITLE
Issues observed during prettifying

### ISSUE THREAD

**Issue by tscholak** [lelit/pglast#55]
Hi, first of all thank you very much for this library. This code is extremely useful, and I couldn't be more happy to have found it.

I have a fairly large collection of SQL queries that I'd like to work on. And while processing the queries, I have found several examples that trigger the "safety belt":

```
SELECT DISTINCT T1.player_name FROM Player AS T1 JOIN Player_Attributes AS T2 ON T1.player_api_id = T2.player_api_id WHERE T2.crossing  >  90 AND T2.preferred_foot  =  "right"
```
becomes
```
SELECT DISTINCT t1.player_name
FROM player AS t1
     INNER JOIN player_attributes AS t2 ON t1.player_api_id = t2.player_api_id
WHERE (t2.crossing > 90)
  AND (t2.preferred_foot = right)
```
Note that `right` is not in quotations marks.

```
SELECT DISTINCT T1.player_name FROM Player AS T1 JOIN Player_Attributes AS T2 ON T1.player_api_id = T2.player_api_id WHERE T2.preferred_foot  =  "left" AND T2.overall_rating  >=  85 AND T2.overall_rating  <=  90
```
becomes
```
SELECT DISTINCT t1.player_name
FROM player AS t1
     INNER JOIN player_attributes AS t2 ON t1.player_api_id = t2.player_api_id
WHERE (t2.preferred_foot = left)
  AND (t2.overall_rating >= 85)
  AND (t2.overall_rating <= 90)
```
Note that `left` is not in quotation marks.

```
SELECT player_api_id FROM Player WHERE height  >=  180 AND height  <=  190 INTERSECT SELECT player_api_id FROM Player_Attributes WHERE preferred_foot  =  "left"
```
becomes
```
SELECT player_api_id
FROM player
WHERE (height >= 180)
  AND (height <= 190)

INTERSECT

SELECT player_api_id
FROM player_attributes
WHERE preferred_foot = left
```
Same problem with `left` and perhaps an issue with `INTERSECT`. Parentheses around the select statements could help.

```
(SELECT document_name FROM documents GROUP BY document_type_code ORDER BY count(*) DESC LIMIT 3) INTERSECT SELECT document_name FROM documents GROUP BY document_structure_code ORDER BY count(*) DESC LIMIT 3
```
becomes
```
SELECT document_name
FROM documents
GROUP BY document_type_code
ORDER BY count(*) DESC
LIMIT 3

INTERSECT

SELECT document_name
FROM documents
GROUP BY document_structure_code
ORDER BY count(*) DESC
LIMIT 3
```
Parentheses around the select statements are mandatory here it would seem.

```
(SELECT document_name FROM documents GROUP BY document_type_code ORDER BY count(*) DESC LIMIT 3) INTERSECT SELECT document_name FROM documents GROUP BY document_structure_code ORDER BY count(*) DESC LIMIT 3
```
becomes
```
SELECT document_name
FROM documents
GROUP BY document_type_code
ORDER BY count(*) DESC
LIMIT 3

INTERSECT

SELECT document_name
FROM documents
GROUP BY document_structure_code
ORDER BY count(*) DESC
LIMIT 3
```
Likewise.

I'm also wondering what happens if one has nested `UNION`, `INTERSECT`, and `EXCEPT` queries, for instance, ```A INTERSECT (B UNION C)``` vs ```(A INTERSECT B) UNION C``` where `A`, `B`, and `C` are placeholders for select statements. It seems it's always a good idea to print parentheses when using set operators.

**Comment by lelit** [comment]
> I have found several examples that trigger the "safety belt"

Thanks for the report: I will try to address all of them, as soon as I'll restart working on v2, hopefully next week.

**Comment by lelit** [comment]
@rdunklau, the ``right`` vs ``"right"`` issue appears to require the list of *reserved words*: maybe there's a way in libpgquery to expose the list defined in ``src/include/parser/kwlist.h``? I could not spend enough time figuring out how the are actually stored, see [keywords.c](https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob_plain;f=src/common/keywords.c;hb=HEAD)

**Comment by rdunklau** [comment]
@lelit I'm not sure I understand, we already have them here https://github.com/lelit/pglast/blob/v2/pglast/keywords.py don't we ?

**Comment by lelit** [comment]
Yes, sorry, one should never read (his own!) source code early in the morning, before first coffee!

**Comment by lelit** [comment]
That problem should be fixable by considering also `TYPE_FUNC_NAME_KEYWORDS`, instead of only `RESERVED_KEYWORDS` [here](https://github.com/lelit/pglast/blob/master/pglast/printer.py#L336). We should add a test to verify the behaviour with all different *kinds* of keyword (that is, checking also some `COL_NAME_KEYWORDS` and `UNRESERVED_KEYWORDS`...)

**Comment by rdunklau** [comment]
I'm not familiar enough with the parser to know what it implies for a keyword to be in one particular set or another. It would seem like depending on the context, you should check some sets and not others...

**Comment by tscholak** [comment]
Thank you very much for the quick response!
I was not aware of the v2 effort. Are you saying that these issues are already fixed in the v2 branch?

Btw, the queries I am considering are from an open source dataset called Spider, https://yale-lily.github.io/spider. There are almost 10k queries in this dataset, and I think it could be a great asset for your work if you wanted to run a lot of there-and-back-again tests.

**Comment by lelit** [comment]
> I was not aware of the v2 effort. Are you saying that these issues are already fixed in the v2 branch?

No, v1 and v2 are very close in functionality, but the latter targets PG12.

> Btw, the queries I am considering are from an open source dataset called Spider, https://yale-lily.github.io/spider. There are almost 10k queries in this dataset, and I think it could be a great asset for your work if you wanted to run a lot of there-and-back-again tests.

Interesting indeed, thanks.

**Comment by lelit** [comment]
@tscholak, I think I cured both problems with the recent commits, any chance you can try current master and (possibly) confirm?

Note to myself: I should spend a tad of time learning the new "GH Actions"...

**Comment by tscholak** [comment]
Hi @lelit, thank you very much for looking into this. The two changes you made fixed the examples reported above, well done! I'm very happy :)

What remained an issue though are nested intersect, union, and except blocks as mentioned in the last paragraph of my initial report.

Can you have a look please at situations like these:
```sql
(
    SELECT * FROM "table"
) INTERSECT (
    (
        SELECT * FROM "table"
    ) UNION (
        SELECT * FROM "table"
    )
)
```
```sql
(
    (
        SELECT * FROM "table"
    ) UNION (
        SELECT * FROM "table"
    )
) INTERSECT (
    SELECT * FROM "table"
)
```
etc.?

**Comment by lelit** [comment]
Thanks, I missed that, sorry. Can you try again with latest master please?

**Comment by tscholak** [comment]
tried it, works as expected!!!

Thanks so much! Consider this issue closed.

**Comment by lelit** [comment]
Great, released as v1.12.
